// phprocess/index.js
// for now just expose the builtin exports global from node.js
//module.exports = global.exports;
// shim for using exports in browser

exports.nextTick = (function () {
    var canSetImmediate = typeof window !== 'undefined'
    && window.setImmediate;
    var canMutationObserver = typeof window !== 'undefined'
    && window.MutationObserver;
    var canPost = typeof window !== 'undefined'
    && window.postMessage && window.addEventListener
    ;

    if (canSetImmediate) {
        return function (f) { return window.setImmediate(f) };
    }

    var queue = [];

    if (canMutationObserver) {
        var hiddenDiv = document.createElement("div");
        var observer = new MutationObserver(function () {
            var queueList = queue.slice();
            queue.length = 0;
            queueList.forEach(function (fn) {
                fn();
            });
        });

        observer.observe(hiddenDiv, { attributes: true });

        return function nextTick(fn) {
            if (!queue.length) {
                hiddenDiv.setAttribute('yes', 'no');
            }
            queue.push(fn);
        };
    }

    if (canPost) {
        window.addEventListener('message', function (ev) {
            var source = ev.source;
            if ((source === window || source === null) && ev.data === 'process-tick') {
                ev.stopPropagation();
                if (queue.length > 0) {
                    var fn = queue.shift();
                    fn();
                }
            }
        }, true);

        return function nextTick(fn) {
            queue.push(fn);
            window.postMessage('process-tick', '*');
        };
    }

    return function nextTick(fn) {
        setTimeout(fn, 0);
    };
})();

exports.title = 'browser';
exports.browser = true;
exports.env = {};
exports.argv = [];

function noop() {}

exports.on = noop;
exports.addListener = noop;
exports.once = noop;
exports.off = noop;
exports.removeListener = noop;
exports.removeAllListeners = noop;
exports.emit = noop;

exports.binding = function (name) {
    throw new Error('exports.binding is not supported');
};

// TODO(shtylman)
exports.cwd = function () { return './' };
exports.chdir = function (dir) {
    throw new Error('exports.chdir is not supported');
};
exports.umask = function() { return 0; };

exports.platform = sys.os.architecture=='32bit' && sys.os.name=='windows' ? 'win32' : 'darwin';

var seprator = isWindows ? '\\' : '/';

exports.getScriptPath = function(  ){
    var script, scriptPath;
    script = exports.system.args[0];
    scriptPath = script.replace(/\/[^\/]+$/, '') // removes everything from 
    //                                              the last "/" until the end of
    //                                              the line, non-greedy
    return scriptPath + seprator;
}
